File: hoffman-agents-node/src/world/self-world.js
const { WorldState } = require('../agent/world-state');

const DEFAULT_GET_STATE = (agent) => ({
  sp: agent.experience.selfToken.stationaryProb,
  pe: agent.meanPredictionError,
  lock: agent.isILocked ? 1 : 0,
  loop: agent.loopScore,
  metaStates: agent.experience.metaTrie.registrySize,
  vocabSize: agent.experience.lexicon.vocabularySize(),
  trieNodes: agent.experience.trie.size(),
});

class SelfWorld {
  constructor(world, agent, getStateFn = null) {
    if (!world || typeof world.step !== 'function') {
      throw new Error('SelfWorld requires a world with a step() method');
    }
    this._inner = world;
    this._agent = agent;
    this._getState = getStateFn || DEFAULT_GET_STATE;
    this._stepCount = 0;
    this.nStates = world.nStates || 1000000;
  }

  step() {
    const worldState = this._inner.step();
    const selfMetrics = this._getState(this._agent);

    const tokens = [];
    for (const [key, val] of Object.entries(selfMetrics)) {
      const str = typeof val === 'number'
        ? `${key}:${Number(val.toFixed(2))}`
        : `${key}:${String(val)}`;
      tokens.push(str);
    }

    worldState.sequences['self'] = tokens;
    worldState.dimensions = { ...worldState.dimensions, ...selfMetrics };

    return worldState;
  }

  get initialState() { return this.step(); }
}

module.exports = { SelfWorld };

File: hoffman-agents-python/src/conscious_agent/world/self_world.py
from __future__ import annotations

from typing import Any, Callable

from ..agent import ConsciousAgent, WorldState


def _default_get_state(agent: ConsciousAgent) -> dict[str, float]:
    return {
        "sp": agent.experience.self_token.stationary_prob,
        "pe": agent.mean_prediction_error,
        "lock": 1.0 if agent.is_i_locked else 0.0,
        "loop": agent.loop_score,
        "meta_states": float(agent.experience.meta_trie.registry_size),
        "vocab_size": float(agent.experience.lexicon.vocabulary_size()),
        "trie_nodes": float(agent.experience.trie.size()),
    }


class SelfWorld:
    def __init__(
        self,
        world: Any,
        agent: ConsciousAgent,
        get_state_fn: Callable[[ConsciousAgent], dict[str, float]] | None = None,
    ) -> None:
        if not hasattr(world, "step") or not callable(world.step):
            raise ValueError("SelfWorld requires a world with a step() method")
        self._inner = world
        self._agent = agent
        self._get_state = get_state_fn or _default_get_state
        self._step_count = 0
        self.n_states = getattr(world, "n_states", 1_000_000)

    def step(self) -> WorldState:
        world_state = self._inner.step()
        self_metrics = self._get_state(self._agent)

        tokens = []
        for key, val in self_metrics.items():
            if isinstance(val, float):
                tokens.append(f"{key}:{val:.2f}")
            else:
                tokens.append(f"{key}:{val}")

        world_state.sequences["self"] = tokens
        if hasattr(world_state, "dimensions"):
            world_state.dimensions = {**world_state.dimensions, **self_metrics}
        else:
            world_state.dimensions = dict(self_metrics)

        return world_state

    @property
    def initial_state(self) -> WorldState:
        return self.step()
